### PR TITLE
add Group Relative Reward Rescaling as length penalty

### PR DESCRIPTION
**Pull request by rasdani**
todo: ablations

---

> [!NOTE]
> **Medium Risk**
> Changes RL advantage/reward shaping by introducing optional length-based GR³ scaling, which can materially affect training dynamics and requires new config validation. Low implementation complexity, but behavioral impact on learning and buffer filtering expectations makes it moderately risky.
> 
> **Overview**
> Adds optional *Group Relative Reward Rescaling (GR³)* length shaping to the default advantage via new config `orchestrator.advantage.length_shaping_alpha`, applying a length-based penalty to rewards before computing the per-problem mean baseline.
> 
> Removes the old `orchestrator.advantage.length_weighted_mean` option, updates config validation to require `buffer.online_difficulty_filtering=true` when GR³ shaping is enabled, and refreshes unit tests and the changelog accordingly.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a6b86a035bd84bc0d8aca3d3888aa92d02fbbd30. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/orchestrator/advantage.py [hunk 1]
```diff
@@ -22,7 +22,13 @@ def compute_advantages(
         return rewards
     rewards = torch.tensor(rewards).view(-1, samples_per_problem)
     lengths = torch.tensor(completion_lengths).view(-1, samples_per_problem)
-    if advantage_config.length_weighted_mean:
+    if advantage_config.gr3_alpha:
+        lengths_normalized = lengths / lengths.mean(dim=1, keepdim=True)
+        length_shaping = (1 + advantage_config.gr3_alpha * lengths_normalized)**-1
+        rewards_corrected = rewards * length_shaping
+        baseline = rewards.mean(dim=1, keepdim=True)
```
[the remark this answers is not in the document]
Comment (rasdani) [reply src/prime_rl/orchestrator/advantage.py]: @claude fix this

File: src/prime_rl/configs/orchestrator.py [hunk 2]
```diff
@@ -552,6 +552,16 @@ class DefaultAdvantageConfig(BaseModel):
 
     type: Literal["default"] = "default"
     length_weighted_mean: bool = False
+    gr3_alpha: Annotated[
+        float | None,
+        Field(description="Penalty coefficient for Group Relative Reward Rescaling. Recommended value: 0.33"),
+    ] = None
```
Comment (samsja) [inline src/prime_rl/configs/orchestrator.py]: can we call it differently

File: src/prime_rl/orchestrator/advantage.py [hunk 3]
```diff
@@ -33,16 +33,26 @@ def my_advantage(inputs: AdvantageInputs, **kwargs) -> AdvantageOutputs:
 """
 
 
-def default_advantage_fn(inputs: AdvantageInputs, length_weighted_mean: bool = False) -> AdvantageOutputs:
+def default_advantage_fn(
+    inputs: AdvantageInputs,
+    length_weighted_mean: bool = False,
+    gr3_alpha: float | None = None,
```
Comment (samsja) [inline src/prime_rl/orchestrator/advantage.py]: so length_weighted_mean has no impact in gr3_alpha ?

Comment (rasdani) [reply src/prime_rl/orchestrator/advantage.py]: yes auto validate makes both configs mutually exclusive.
they are two distinct forms of length penalty.

File: src/prime_rl/configs/orchestrator.py (hunk 2 above)
Comment (rasdani) [reply src/prime_rl/configs/orchestrator.py]: renamed to `length_shaping_alpha`

**Review by samsja — APPROVED** [review]
